File: brain/singleton.py
import atexit
import os
import signal
from pathlib import Path
from typing import Callable, Optional


def _is_process_alive(pid: int) -> bool:
    if pid <= 0:
        return False

    try:
        os.kill(pid, 0)
        return True
    except OSError:
        return False


def _read_existing_pid(lock_path: Path) -> Optional[int]:
    try:
        raw = lock_path.read_text(encoding="utf-8").strip()
        return int(raw)
    except (OSError, ValueError):
        return None


def acquire_process_lock(lock_name: str) -> Optional[Callable[[], None]]:
    lock_dir = Path(__file__).resolve().parent.parent / "data" / "locks"
    lock_dir.mkdir(parents=True, exist_ok=True)

    safe_name = "".join(ch if ch.isalnum() or ch in {"_", "-"} else "_" for ch in lock_name).lower()
    lock_path = lock_dir / f"{safe_name}.lock"

    try:
        fd = os.open(str(lock_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY)
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            handle.write(f"{os.getpid()}\n")
    except FileExistsError:
        existing_pid = _read_existing_pid(lock_path)
        if existing_pid and _is_process_alive(existing_pid):
            print(
                f"[LOCK] {lock_name} is already running under PID {existing_pid}. "
                "This duplicate instance will exit."
            )
            return None

        try:
            lock_path.unlink()
        except OSError:
            print(f"[LOCK] Could not remove stale lock for {lock_name} at {lock_path}.")
            return None

        return acquire_process_lock(lock_name)

    released = False

    def release() -> None:
        nonlocal released
        if released:
            return
        released = True
        try:
            if lock_path.exists():
                lock_path.unlink()
        except OSError:
            pass

    atexit.register(release)

    def _handle_signal(signum, frame):
        release()
        raise SystemExit(0)

    for sig_name in ("SIGINT", "SIGTERM"):
        sig = getattr(signal, sig_name, None)
        if sig is not None:
            signal.signal(sig, _handle_signal)

    return release

File: app/singleton.ts
import * as fs from "fs";
import * as path from "path";

type ReleaseFn = () => void;

function isProcessAlive(pid: number): boolean {
  if (!Number.isInteger(pid) || pid <= 0) {
    return false;
  }

  try {
    process.kill(pid, 0);
    return true;
  } catch (e: any) {
    return e?.code === "EPERM";
  }
}

function readExistingPid(lockPath: string): number | null {
  try {
    const raw = fs.readFileSync(lockPath, "utf-8").trim();
    const parsed = Number.parseInt(raw, 10);
    return Number.isInteger(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

export function acquireProcessLock(lockName: string): ReleaseFn | null {
  const lockDir = path.join(__dirname, "../data/locks");
  fs.mkdirSync(lockDir, { recursive: true });

  const safeName = lockName.replace(/[^a-z0-9_-]/gi, "_").toLowerCase();
  const lockPath = path.join(lockDir, `${safeName}.lock`);

  try {
    const fd = fs.openSync(lockPath, "wx");
    fs.writeFileSync(fd, `${process.pid}\n`, "utf-8");
    fs.closeSync(fd);
  } catch (e: any) {
    if (e?.code !== "EEXIST") {
      throw e;
    }

    const existingPid = readExistingPid(lockPath);
    if (existingPid && isProcessAlive(existingPid)) {
      console.error(
        `[LOCK] ${lockName} is already running under PID ${existingPid}. ` +
        `This duplicate instance will exit.`
      );
      return null;
    }

    try {
      fs.unlinkSync(lockPath);
    } catch {
      console.error(`[LOCK] Could not remove stale lock for ${lockName} at ${lockPath}.`);
      return null;
    }

    return acquireProcessLock(lockName);
  }

  let released = false;
  const release = () => {
    if (released) return;
    released = true;
    try {
      if (fs.existsSync(lockPath)) {
        fs.unlinkSync(lockPath);
      }
    } catch {
      // Best-effort cleanup during shutdown.
    }
  };

  process.on("exit", release);
  process.on("SIGINT", () => {
    release();
    process.exit(0);
  });
  process.on("SIGTERM", () => {
    release();
    process.exit(0);
  });
  process.on("uncaughtException", (err) => {
    release();
    throw err;
  });

  return release;
}
